android/app/src/main/java/com/zulipmobile/MainActivity.kt

```diff
@@ -7,12 +7,31 @@ import android.webkit.WebView
 import com.facebook.react.ReactActivity
 import com.facebook.react.ReactActivityDelegate
 import com.facebook.react.ReactApplication
+import com.facebook.react.ReactInstanceManager
+import com.facebook.react.ReactNativeHost
 import com.facebook.react.ReactRootView
 import com.facebook.react.bridge.ReactContext
+import com.facebook.react.modules.core.DeviceEventManagerModule
 import com.zulipmobile.notifications.*
 import com.zulipmobile.sharing.handleSend
 import expo.modules.ReactActivityDelegateWrapper
 
+// A convenience shortcut.
+fun ReactContext.emit(eventName: String, data: Any?) {
+    getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
+        .emit(eventName, data)
+}
+
+/**
+ * Like getReactInstanceManager, but just return what exists; avoid trying to create.
+ *
+ * When there isn't already an instance manager, if we call
+ * getReactInstanceManager it'll try to create one... which asserts we're
+ * on the UI thread, which isn't true if e.g. we got here from a Service.
+ */
+fun ReactNativeHost.tryGetReactInstanceManager(): ReactInstanceManager? =
+    if (this.hasInstance()) this.reactInstanceManager else null
+
 // A convenience shortcut.
 fun ReactApplication.tryGetReactContext(): ReactContext? =
```

Comment:
Hmm, seeing all these together: let's find them a place to live that isn't "MainActivity.kt". They're really not about the main activity; a couple of them lived here because they were helpers used by the main activity's implementation, but that rationale no longer works.

I think `ReactExtensions.kt`, in this main com/zulipmobile/ directory, would be a good name. Then the ReactAppStatus stuff can go there too.

According to [the upstream Kotlin style guide](https://kotlinlang.org/docs/coding-conventions.html#source-code-organization), these would go in the same files where `ReactNativeHost`, `ReactApplication`, and so on are defined. That's not an option, so I think a `ReactExtensions.kt` is the next best.